File: apps/desktop/src/lib/mock/board.ts
// Мок канбан-доски (BOARD-2/4): зеркалит контракт Rust `board::list_board` — список заметок-задач
// (frontmatter `status`) с полями для карточек. Сидовый набор покрывает дефолтные колонки
// (todo/doing/done), off-set статус («ожидание» → виртуальная «Прочее»), проекты, приоритеты,
// дедлайны (в т.ч. просроченный) и теги — чтобы превью BoardView было содержательным.

import type { BoardConfig, BoardData, BoardSummary, StaleTask, TaskCard } from '../tauri-api';

const SEED: TaskCard[] = [
  {
    path: 'Tasks/Дизайн доски.md',
    title: 'Дизайн канбан-доски',
    status: 'doing',
    project: 'Nexus',
    priority: 'high',
    due: '2026-06-20',
    tags: ['design', 'task'],
  },
  {
    path: 'Tasks/Перенести заметки.md',
    title: 'Перенести старые заметки',
    status: 'todo',
    project: 'Nexus',
    priority: 'medium',
    due: null,
    tags: ['migration'],
  },
  {
    path: 'Tasks/Оплатить счёт.md',
    title: 'Оплатить счёт за интернет',
    status: 'todo',
    project: 'Дом',
    priority: 'urgent',
    due: '2026-06-14', // просрочено относительно 2026-06-16
    tags: ['быт'],
  },
  {
    path: 'Tasks/Прочитать статью.md',
    title: 'Прочитать статью про RAG',
    status: 'done',
    project: 'Учёба',
    priority: 'low',
    due: null,
    tags: ['reading', 'ai'],
  },
  {
    path: 'Tasks/Согласовать смету.md',
    title: 'Согласовать смету с подрядчиком',
    status: 'ожидание', // вне дефолтного набора → виртуальная колонка «Прочее»
    project: 'Дом',
    priority: 'medium',
    due: '2026-06-25',
    tags: [],
  },
  {
    path: 'Tasks/Релиз 0.9.md',
    title: 'Подготовить релиз 0.9',
    status: 'doing',
    project: 'Nexus',
    priority: 'high',
    due: '2026-06-30',
    tags: ['release', 'task'],
  },
];

/** Все заметки-задачи с полями для доски (сид под дефолтный ключ `status`). */
export async function listBoard(): Promise<TaskCard[]> {
  // Возвращаем копию, чтобы потребитель не мутировал сид.
  return SEED.map((c) => ({ ...c, tags: [...c.tags] }));
}

// BOARD-3: персист-конфиг доски — мутабельный мок (saveBoard переживает в рамках сессии, как бэкенд-файл).
let CONFIG: BoardConfig = {
  id: 'personal',
  title: '',
  statusKey: 'status',
  columns: [
    { id: 'todo', label: '', wip: null, color: null, doneLike: false },
    { id: 'doing', label: '', wip: null, color: null, doneLike: false },
    { id: 'done', label: '', wip: null, color: null, doneLike: true },
  ],
  scope: { folder: null, project: null, tags: [] },
  order: { doing: ['Tasks/Релиз 0.9.md', 'Tasks/Дизайн доски.md'] }, // показываем ручной порядок в превью
  sort: 'manual',
  cardFields: ['due', 'priority', 'tags'],
};

/** Доска целиком (зеркало `get_board`): конфиг + карточки в scope + corrupt. Мок — scope пустой → все. */
export async function getBoard(): Promise<BoardData> {
  return {
    config: structuredClone(CONFIG),
    cards: SEED.map((c) => ({ ...c, tags: [...c.tags] })),
    corrupt: false,
  };
}

/** Персист конфига (зеркало `save_board`) — мутирует мок-стейт. */
export async function saveBoard(config: BoardConfig): Promise<void> {
  CONFIG = structuredClone(config);
}

/** Список досок (зеркало `list_boards`) — всегда ≥1. */
export async function listBoards(): Promise<BoardSummary[]> {
  return [{ id: CONFIG.id, title: CONFIG.title }];
}

/** AI-2a: «застрявшие» задачи (зеркало `stale_tasks`). Сид включает done-задачу — фронт ОБЯЗАН её
 *  отсеять (done-like колонка), демонстрируя контракт фильтрации; статусы зеркалят SEED. */
export async function staleTasks(): Promise<StaleTask[]> {
  return [
    {
      path: 'Tasks/Согласовать смету.md',
      title: 'Согласовать смету с подрядчиком',
      status: 'ожидание',
      lastEdit: 1_747_000_000,
      daysStale: 42,
    },
    {
      path: 'Tasks/Перенести заметки.md',
      title: 'Перенести старые заметки',
      status: 'todo',
      lastEdit: 1_748_000_000,
      daysStale: 21,
    },
    {
      // done-задача застряла по времени, но фронт её НЕ покажет (done-like).
      path: 'Tasks/Прочитать статью.md',
      title: 'Прочитать статью про RAG',
      status: 'done',
      lastEdit: 1_746_000_000,
      daysStale: 50,
    },
  ];
}

File: apps/desktop/src-tauri/src/commands/board.rs
//! Команды канбан-доски: BOARD-2 (выборка задач) + BOARD-3 (персист конфига колонок/порядка/scope).

use std::collections::HashSet;

use serde::Serialize;
use tauri::State;

use crate::board::config::{self, BoardConfig, BoardSummary};
use crate::board::{self, StaleTask, TaskCard, DEFAULT_STATUS_KEY};
use crate::error::{AppError, AppResult};
use crate::state::AppState;

/// Порог «застряло» по умолчанию (дней) — AI-2a; фронт может переопределить.
const DEFAULT_STALE_DAYS: i64 = 14;

/// Все заметки-задачи (есть frontmatter-ключ `status_key`, по умолч. `status`) с полями для доски.
/// Без открытого vault — ошибка. Чистый SQL-read (офлайн, без LLM/сети). Колонкование — на фронте.
#[tauri::command]
pub async fn list_board(
    state: State<'_, AppState>,
    status_key: Option<String>,
) -> AppResult<Vec<TaskCard>> {
    let reader = state.vault().await?.db.reader().clone();
    let key = status_key
        .filter(|k| !k.trim().is_empty())
        .unwrap_or_else(|| DEFAULT_STATUS_KEY.to_string());
    Ok(board::list_board(&reader, key).await?)
}

/// AI-2a (A2): «застрявшие» задачи — заметки-задачи, не правленные ≥ `threshold_days` (умолч. 14) дней по
/// `edit_events` (фолбэк mtime). Детерминированный SQL-read (без LLM/сети). `now` — серверное время.
/// Done-like-статусы НЕ отсеиваются здесь (фронт фильтрует по конфигу доски).
#[tauri::command]
pub async fn stale_tasks(
    state: State<'_, AppState>,
    status_key: Option<String>,
    threshold_days: Option<i64>,
) -> AppResult<Vec<StaleTask>> {
    let reader = state.vault().await?.db.reader().clone();
    let key = status_key
        .filter(|k| !k.trim().is_empty())
        .unwrap_or_else(|| DEFAULT_STATUS_KEY.to_string());
    let days = threshold_days.unwrap_or(DEFAULT_STALE_DAYS).max(1);
    Ok(board::stale_tasks(&reader, key, days, crate::scheduler::now_secs()).await?)
}

/// Доска целиком (BOARD-3): персист-конфиг (колонки/порядок/scope) + карточки в его scope/statusKey.
/// `corrupt` — JSON конфига был, но битый → фронт-тост (используется дефолт, файл НЕ перезаписан вслепую).
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BoardData {
    pub config: BoardConfig,
    pub cards: Vec<TaskCard>,
    pub corrupt: bool,
}

/// Загружает доску `slug` (умолч. `personal`): конфиг + карточки. order самозалечивается — GC удалённых/
/// вне-scope путей; при изменении и ВАЛИДНОМ конфиге best-effort persist (битый не перезаписываем).
#[tauri::command]
pub async fn get_board(state: State<'_, AppState>, slug: Option<String>) -> AppResult<BoardData> {
    let ctx = state.vault().await?;
    let root = ctx.root.clone();
    let reader = ctx.db.reader().clone();
    let id = slug
        .filter(|s| !s.trim().is_empty())
        .unwrap_or_else(|| config::BOARD_ID_DEFAULT.to_string());

    let root_l = root.clone();
    let id_l = id.clone();
    let loaded = tokio::task::spawn_blocking(move || config::load(&root_l, &id_l))
        .await
        .map_err(|e| AppError::Msg(e.to_string()))?;
    let mut cfg = loaded.config;

    let all = board::list_board(&reader, cfg.status_key.clone()).await?;
    let cards: Vec<TaskCard> = all
        .into_iter()
        .filter(|c| config::matches_scope(c, &cfg.scope))
        .collect();

    // GC порядка — ТОЛЬКО для отображения (in-memory): убираем из возвращаемого order пути вне текущих
    // карточек. НЕ персистим на чтении! (adversarial-ревью F1, класс MEM-5): при холодном/отстающем
    // индексе `list_board` отдаёт меньше карточек, чем есть на диске → персист GC СТЁР БЫ ручной порядок
    // живых задач. Чистим персист ТОЧЕЧНО на реальном удалении (delete-хук в `delete_path`) и реордере
    // (`save_board`), а не вслепую по «отсутствию в выборке». Стейл-записи в файле безвредны (applyOrder
    // их игнорирует).
    let existing: HashSet<&str> = cards.iter().map(|c| c.path.as_str()).collect();
    config::gc_order(&mut cfg, &existing);

    Ok(BoardData {
        config: cfg,
        cards,
        corrupt: loaded.corrupt,
    })
}

/// Персистит конфиг доски (BOARD-3): переименование/реордер колонок, ручной порядок (DnD — BOARD-5).
#[tauri::command]
pub async fn save_board(state: State<'_, AppState>, config: BoardConfig) -> AppResult<()> {
    let root = state.vault().await?.root.clone();
    tokio::task::spawn_blocking(move || config::save(&root, &config))
        .await
        .map_err(|e| AppError::Msg(e.to_string()))?
        .map_err(AppError::Io)?;
    Ok(())
}

/// Список досок (`.nexus/boards/*.json`); пусто → синтетический дефолт (всегда ≥1 доска для UI).
#[tauri::command]
pub async fn list_boards(state: State<'_, AppState>) -> AppResult<Vec<BoardSummary>> {
    let root = state.vault().await?.root.clone();
    tokio::task::spawn_blocking(move || config::list_boards(&root))
        .await
        .map_err(|e| AppError::Msg(e.to_string()))
}
